<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Recycler : This is the codebase for Recycler, the new tool described in our manuscript: http://biorxiv.org/content/early/2016/04/05/029926">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Recycler</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/Shamir-Lab/Recycler">View on GitHub</a>

          <h1 id="project_title">Recycler</h1>
          <h2 id="project_tagline">This is the codebase for Recycler, the new tool described in our manuscript: http://biorxiv.org/content/early/2016/04/05/029926</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/Shamir-Lab/Recycler/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/Shamir-Lab/Recycler/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>
<a id="getting-recycler" class="anchor" href="#getting-recycler" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Getting Recycler</h1>

<p>You can download Recycler <a href="https://github.com/Shamir-Lab/Recycler/releases/download/Recycler-v0.6/Recycler-0.6.zip">here</a> or clone it via the link below. In case you download the zip, unzip the file before following the instructions below (ignoring the 'git clone' line)</p>

<h1>
<a id="quick-start" class="anchor" href="#quick-start" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Quick start</h1>

<p>Assuming we have prepared a filtered BAM file (aln-pe.bam) prepared as described <a href="#bam-prep">below</a> and an isolate assembly graph (e.g., assembly_graph.fastg from <a href="http://bioinf.spbau.ru/en/spades">SPAdes 3.6+</a>), and that 55 was the maximum k-mer length used by the assembler, </p>

<pre><code>git clone https://github.com/rozovr/Recycler.git
cd Recycler
python recycle.py -g assembly_graph.fastg -k 55 -b aln-pe.bam -i True
</code></pre>

<p>For metagenome/plasmidome assemblies, we remove the final ("-i") parameter, which has a default False value.</p>

<h1>
<a id="introduction" class="anchor" href="#introduction" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Introduction</h1>

<p>Recycler is a tool designed for extracting circular sequences from de novo assembly graphs. It can be applied on isolate as well as metagenome and plasmidome data. The circular sequences it outputs may be plasmids, phages, etc. Recycler uses only features of the assembly graph (contig overlaps, lengths, and coverage level information)  and alignments of paired-end reads to the contigs in the graph in order to identify these sequences.  </p>

<h1>
<a id="requirements" class="anchor" href="#requirements" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Requirements</h1>

<p>Recycler is implemented in Python, and has been tested only on Python 2.7+. We recommend using the Anaconda distribution to ease package installations.
Recycler requires the following packages be installed:</p>

<ul>
<li><a href="http://www.numpy.org/">NumPy</a></li>
<li><a href="http://networkx.github.io/">NetworkX</a></li>
<li><a href="https://github.com/pysam-developers/pysam">PySAM</a></li>
<li><a href="https://nose.readthedocs.org/en/latest/">nose</a></li>
</ul>

<p>Recommended for generating inputs (as used during testing):</p>

<ul>
<li><a href="https://github.com/lh3/bwa">BWA 0.7.5+</a></li>
<li><a href="https://github.com/samtools/samtools">samtools 0.1.19+</a></li>
<li><a href="http://bioinf.spbau.ru/en/spades">SPAdes 3.6.2+</a></li>
</ul>

<h1>
<a id="detailed-usage" class="anchor" href="#detailed-usage" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Detailed usage</h1>

<p>python recycle.py -g GRAPH -k MAX_K -b BAM [-l LENGTH] [-m MAX_CV] [-i ISO]</p>

<h3>
<a id="required-arguments" class="anchor" href="#required-arguments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>required arguments:</h3>

<pre><code>-g GRAPH
(spades 3.50+) assembly graph FASTG file to process:
for spades 3.5, before_rr.fastg; for spades 3.6+, assembly_graph.fastg
-k MAX_K
integer reflecting maximum k value used by the assembler
-b BAM
BAM file resulting from aligning reads to contigs file, filtering for best matches
</code></pre>

<h3>
<a id="optional-arguments" class="anchor" href="#optional-arguments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>optional arguments:</h3>

<pre><code>-l LENGTH
minimum length required for reporting [default: 1000]
-m MAX_CV
coefficient of variation used for pre-selection
[default: 0.5, higher--&gt; less restrictive]
-i ISO
True or False value reflecting whether data sequenced
was an isolated strain 
</code></pre>

<h1>
<a id="preparing-the-bam-input" class="anchor" href="#preparing-the-bam-input" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><a name="bam-prep">Preparing the BAM input:</a>
</h1>

<p>Recycler uses paired-end alignments of the reads originally assembled to the output assembly graph to filter and select amongst candidate circular sequences. In order to do so, it requires as input a BAM file containing the set of best alignment hits for each read pair. We recommend the following steps (tested on BWA 0.7.5 and samtools 1.19) to prepare the BAM file:</p>

<pre><code>python make_fasta_from_fastg.py -g assembly_graph.fastg

bwa index assembly_graph.nodes.fasta

bwa mem  assembly_graph.nodes.fasta R1.fastq.gz R2.fastq.gz | samtools view -buS - &gt; reads_pe.bam

samtools view -bF 0x0800 reads_pe.bam &gt; reads_pe_primary.bam

samtools sort reads_pe_primary.bam reads_pe_primary.sort

samtools index reads_pe_primary.sort.bam
</code></pre>

<p>following these steps, we only need the files reads_pe_primary.sort.bam and reads_pe_primary.sort.bam.bai.</p>

<h1>
<a id="outputs" class="anchor" href="#outputs" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Outputs:</h1>

<ol>
<li>&lt;prefix&gt;.cycs.fasta  - a fasta formatted file of predicted plasmids</li>
<li>&lt;prefix&gt;.cycs.paths_w_cov.txt - a text file containing information about plasmids composed of multiple contigs.</li>
</ol>

<p>The format for the second file is:</p>

<ul>
<li>
<em>&lt;plasmid name&gt;</em> - e.g., RNODE_5_length_42666_cov_19.93685</li>
<li>
<em>&lt;node names in the original graph making up this cycle&gt;</em> - e.g., ('NODE_2801_length_42596_cov_19.8677', "NODE_2387_length_125_cov_34.7286'").</li>
<li>
<em>&lt;coverage levels of nodes at the time they are removed&gt;</em> - e.g., [19.8677, 34.7286]</li>
<li>
<em>&lt;node numbers in the original graph making up this cycle&gt;</em> - e.g., [2801, 2387]. This can be useful for visualizing the path in tools like <a href="https://rrwick.github.io/Bandage/">Bandage</a>
</li>
</ul>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Recycler maintained by <a href="https://github.com/Shamir-Lab">Shamir-Lab</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
